File: src/main/com/natpryce/snodge/mutagens/RemoveJsonElement.java
package com.natpryce.snodge.mutagens;

import com.google.gson.JsonElement;
import com.natpryce.snodge.DocumentMutation;
import com.natpryce.snodge.JsonPath;
import com.natpryce.snodge.Mutagen;
import kotlin.sequences.Sequence;
import kotlin.sequences.SequencesKt;

public class RemoveJsonElement implements Mutagen {
    @Override
    public Sequence<DocumentMutation> potentialMutations(JsonElement document, final JsonPath pathToElement, JsonElement elementToMutate) {
        if (pathToElement.isRoot()) {
            return SequencesKt.emptySequence();
        } else {
            return SequencesKt.sequenceOf(pathToElement.remove());
        }
    }
}

File: src/main/com/natpryce/snodge/mutagens/RemoveJsonElement.kt
package com.natpryce.snodge.mutagens

import com.google.gson.JsonElement
import com.natpryce.snodge.JsonNodeMutagen
import com.natpryce.snodge.JsonPath

class RemoveJsonElement : JsonNodeMutagen {
    override fun invoke(document: JsonElement, pathToElement: JsonPath, elementToMutate: JsonElement): Sequence<Lazy<JsonElement>> {
        if (pathToElement.isRoot) {
            return emptySequence()
        }
        else {
            return sequenceOf(lazy { pathToElement.remove(document) })
        }
    }
}
